packages/auth/src/OAuth.ts

```diff
@@ -1,12 +1,166 @@
+import { Popup } from '@salte-auth/popup';
+import { SalteAuth } from '@salte-auth/salte-auth';
+import mitt from 'mitt';
+import { AuthParameters } from './AuthParameters';
+import CartoProvider from './CartoProvider';
+import { CARTO_AUTHORIZATION_BASE, THRESHOLD } from './constants';
+import Credentials from './Credentials';
+import { Iframe } from './Iframe';
+import UserInfo from './UserInfo';
+import { unknownScopes } from './utils';
+
 class OAuth {
-  private _clientId: string;
+  private _client: SalteAuth;
+  private _carto: CartoProvider;
+  private _refresher: Iframe;
+  private _refreshTimeout: number = -1;
+  private _emitter: mitt.Emitter;
+
+  constructor(args: AuthParameters, refreshUrl: string | undefined) {
+    if (args == null) {
+      throw new Error('Missing OAuth parameters');
+    }
+
+    args.authorization = args.authorization || CARTO_AUTHORIZATION_BASE;
+    args.scopes = args.scopes || '';
+
+    this._emitter = new mitt();
+
+    this._client = new SalteAuth({
+      providers: [
+        new CartoProvider({
+          storage: 'local',
+          level: 'trace',
+          url: args.authorization,
+          redirectUrl: args.redirectURI,
+          clientID: args.clientID,
+          responseType: 'token',
+          scope: args.scopes
+        })
+      ],
+
+      handlers: [
+        new Popup({
+          default: true
+        })
+      ]
+    });
+
+    this._refresher = new Iframe({
+      authorization: args.authorization,
+      redirectURI: refreshUrl || args.redirectURI,
+      clientID: args.clientID,
+      scopes: args.scopes
+    });
+
+    // Read the expires_in, setup timer
+    this._carto = this._client.provider('carto') as CartoProvider;
+
+    if (this._carto.expiresIn !== -1) {
+      this._scheduleRefresh();
+    }
+
+    if (args.scopes) {
+      this._checkScopes(args.scopes);
+    }
+  }
+
+  public login() {
+    return new Promise((resolve, reject) => {
+      const cb = (error?: Error, token?: SalteAuth.EventWrapper) => {
+
+        if (error) {
+          reject(error);
+          this._emitter.emit('error');
```

Comment:
Are we keeping this dual approach in other places, reject + event?

Reply:
Well, it's complicated.

On the one hand, we are using an event system because refreshes happen completely asynchronously, and the consumer of the library cannot not have a Promise for that. Or at least, I don't see a nice way.

On the other hand, certain apps will most likely want a login.then().catch() approach. For instance, the one we're building this for.